core/services/functions/orm.go

```diff
@@ -186,11 +189,47 @@ func (o orm) FindOldestEntriesByState(state RequestState, limit uint32, qopts ..
 	return requests, nil
 }
 
-func (o orm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
+func (o *orm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
 	var request Request
 	stmt := fmt.Sprintf(`SELECT %s FROM ocr2dr_requests WHERE request_id=$1 AND contract_address=$2;`, requestFields)
 	if err := o.q.WithOpts(qopts...).Get(&request, stmt, requestID, o.contractAddress); err != nil {
 		return nil, err
 	}
 	return &request, nil
 }
+
+func (o *orm) PruneOldestRequests(maxStoredRequests uint32, batchSize uint32, qopts ...pg.QOpt) (total uint32, pruned uint32, err error) {
+	err = o.q.WithOpts(qopts...).Transaction(func(tx pg.Queryer) error {
+		if err2 := tx.Get(&total, "SELECT COUNT(*) FROM ocr2dr_requests WHERE contract_address=$1", o.contractAddress); err2 != nil {
+			return errors.Wrap(err, "failed to get request count")
+		}
+
+		if total <= maxStoredRequests {
+			pruned = 0
+			return nil
+		}
+
+		pruneLimit := total - maxStoredRequests
+		if pruneLimit > batchSize {
+			pruneLimit = batchSize
+		}
+
+		deleteStmt := `
+		  DELETE FROM ocr2dr_requests WHERE contract_address = $1 AND request_id IN (
```

Comment:
It should be CTE query, see https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-cte/
I have done the same in S4 ORM for limited GC.

Reply:
How can I extract "total" from a CTE query?